### PR TITLE
Renderer update and return frame loop

### PR DESCRIPTION
**Pull request by PaulHax**
Depends on: https://github.com/thewtex/agave-hypha-kubernetes/pull/1

### CODE REVIEW

File: packages/remote-viewport/src/remote-machine.ts [hunk 1]
```diff
@@ -1,11 +1,22 @@
...
+
 import { Viewport } from '@itk-viewer/viewer/viewport.js';
-import { ActorRef, assign, createMachine } from 'xstate';
+
+type RendererProps = {
+  density: number;
+  cameraPose: ReadonlyMat4;
+};
+
+type RendererPatch = Partial<RendererProps>;
```
Comment (thewtex) [inline packages/remote-viewport/src/remote-machine.ts:11]: In order to:

1. Keep the logic of how this is handled on the receiving end simple and clear.
2. Keep the impact of the effect of the order of events consistently applied.

the patch should be an Array of Event's that can be sent to the renderer.

The renderer's implementation for handling these will be a simple for-loop with a switch statement on the event type.

File: packages/remote-viewport/src/remote-machine.ts [hunk 2]
```diff
@@ -78,19 +96,59 @@ export const remoteMachine = createMachine({
...
+              (self as ActorRef<UpdateRendererEvent | RenderEvent>).send({
+                type: 'render',
+              });
+            },
+          ],
+        },
+      },
+      initial: 'preRender',
       states: {
+        // consumes queue in prep for renderer (as "entry" action happens before "invoke:input")
```
Comment (PaulHax) [inline packages/remote-viewport/src/remote-machine.ts:132]: Change to say entry action happens after invoke:input

File: packages/remote-viewport/src/remote-machine.ts [hunk 3]
```diff
@@ -78,19 +96,59 @@ export const remoteMachine = createMachine({
...
+              getEntries(context.rendererProps),
           }),
-          target: 'connected',
+          target: 'rendering',
         },
       },
     },
-    connected: {
-      initial: 'render',
+    rendering: {
```
Comment (PaulHax) [inline packages/remote-viewport/src/remote-machine.ts:107]: change name to online

File: packages/remote-viewport/src/remote-machine.ts [hunk 4]
```diff
@@ -100,35 +158,14 @@ export const remoteMachine = createMachine({
...
-              ],
-              target: 'updating',
-            },
-            cameraPoseUpdated: {
-              target: 'updating',
-            },
+          always: {
+            // Renderer props changed while rendering? Then render.
+            guard: ({ context }) =>
+              Object.keys(context.queuedRendererProps).length > 0,
```
Comment (PaulHax) [inline packages/remote-viewport/src/remote-machine.ts:164]: This is array now, just check length

File: packages/remote-viewport/src/remote-machine.ts [hunk 5]
```diff
@@ -78,19 +96,59 @@ export const remoteMachine = createMachine({
...
+              });
+            },
+          ],
+        },
+      },
+      initial: 'preRender',
       states: {
+        // consumes queue in prep for renderer (as "entry" action happens before "invoke:input")
+        preRender: {
+          entry: assign({
```
Comment (thewtex) [inline packages/remote-viewport/src/remote-machine.ts:134]: instead exit from idle

**Comment by PaulHax** [comment]
I think this is good to merge.

**Review by thewtex — APPROVED** [review]

**Pull request merged**